transport/x/redis/inbound.go

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package redis
+
+import (
+	"context"
+	"time"
+
+	"go.uber.org/yarpc/internal/errors"
+	"go.uber.org/yarpc/internal/request"
+	"go.uber.org/yarpc/serialize"
+	"go.uber.org/yarpc/transport"
+	"go.uber.org/yarpc/transport/internal"
+
+	"github.com/opentracing/opentracing-go"
+)
+
+const transportName = "redis"
+
+// Inbound is a redis inbound that reads from the given queueKey. This will
+// wait for an item in the queue or until the timout is reached before trying
+// to read again.
+type Inbound struct {
+	registry transport.Registry
+	tracer   opentracing.Tracer
+
+	client        Client
+	timeout       time.Duration
+	queueKey      string
+	processingKey string
+
+	stop chan struct{}
+}
+
+// NewInbound creates a redis Inbound that satisfies transport.Inbound.
+//
+// queueKey - key for the queue in redis
+// processingKey - key for the list we'll store items we've popped from the queue
+// timeout - how long the inbound will block on reading from redis
+func NewInbound(client Client, queueKey, processingKey string, timeout time.Duration) *Inbound {
+	return &Inbound{
+		tracer: opentracing.GlobalTracer(),
+
+		client:        client,
+		timeout:       timeout,
+		queueKey:      queueKey,
+		processingKey: processingKey,
+
+		stop: make(chan struct{}),
+	}
+}
+
+// WithTracer configures a tracer on this inbound.
+func (i *Inbound) WithTracer(tracer opentracing.Tracer) *Inbound {
```

Comment:
if you observe the changes made by @kriskowal here: https://github.com/yarpc/yarpc-go/pull/540/files#diff-80e9e3ac4681dfe8425e0cef31b983dbR36
We invert the setting of tracer to be set on the "transport" object itself.  Might make sense to do the same split here.

Since this is in a /x/ package this is non-blocking though.